File: _stats_build/statistics/average_of.rb
# NOTE: 聚合页面——将 5 个 AverageOfX 子类合并为一个入口
# 用户通过选择器在 Ao5 / Ao12 / Ao25 / Ao50 / Ao100 之间切换
require_relative "abstract/average_of_x"
require_relative "../core/stat_panel"
require_relative "average_of_3"
require_relative "../core/metric_layout"
require_relative "average_of_5"
require_relative "average_of_12"
require_relative "average_of_25"
require_relative "average_of_50"
require_relative "average_of_100"

class AverageOf < Statistic
  include StatPanel
  include MetricLayout

  AOX_CLASSES = [AverageOf3, AverageOf5, AverageOf12, AverageOf25, AverageOf50, AverageOf100].freeze

  # NOTE: 按钮标签和 HTML ID 前缀
  AOX_META = {
    AverageOf3   => { label: "Mo3",   id: "mo3" },
    AverageOf5   => { label: "Ao5",   id: "ao5" },
    AverageOf12  => { label: "Ao12",  id: "ao12" },
    AverageOf25  => { label: "Ao25",  id: "ao25" },
    AverageOf50  => { label: "Ao50",  id: "ao50" },
    AverageOf100 => { label: "Ao100", id: "ao100" },
  }.freeze

  def initialize
    @title = "Rolling Average"
    @note = "X consecutive official attempts are considered. " \
            "Top N varies by event: 333 top 15, most events top 30, " \
            "333fm/pyram/clock/444bf/555bf/magic/mmagic top 300, " \
            "333bf top 2000."
  end

  def markdown
    md = top
    md += metric_tab_wrap_start

    md += aggregate_panels(AOX_CLASSES, AOX_META) do |inst, prefix|
      # NOTE: 双 Tab 视图——复用 wr_aoxr.rb 的 grouped_panel 模式
      ranking = inst.ranking_data
      history = inst.wr_history

      panel = grouped_panel("#{prefix}-ranking", true, ranking, AverageOfX::RANKING_HEADER_AOX,
                            label_en: "Current Ranking", label_zh: "排名")
      panel += grouped_panel("#{prefix}-history", false, history, inst.history_header,
                             label_en: "WR History", label_zh: "历史")
      panel
    end

    md += metric_tab_wrap_end
    md
  end
end

File: trainer/packages/stats-build/src/statistics/average_of.ts
// NOTE: Average Of X 聚合页面——将 7 个 AverageOfX 子类合并为一个 JSON
// 与 Ruby _stats_build/statistics/average_of.rb 1:1 对应
import { Statistic } from '../core/statistic.js';
import type { StatJson, MetricPanel } from '../core/statistic.js';

// NOTE: 7 个 AverageOfX 子类定义
const AOX_DEFS = [
  { module: () => import('./average_of_3.js'),    label: 'Ao3',    id: 'ao3' },
  { module: () => import('./average_of_5.js'),    label: 'Ao5',    id: 'ao5' },
  { module: () => import('./average_of_12.js'),   label: 'Ao12',   id: 'ao12' },
  { module: () => import('./average_of_25.js'),   label: 'Ao25',   id: 'ao25' },
  { module: () => import('./average_of_50.js'),   label: 'Ao50',   id: 'ao50' },
  { module: () => import('./average_of_100.js'),  label: 'Ao100',  id: 'ao100' },
  { module: () => import('./average_of_1000.js'), label: 'Ao1000', id: 'ao1000' },
] as const;

export class AverageOf extends Statistic {
  constructor() {
    super();
    this.title = 'Average of X';
    this.titleZh = 'X 次均值';
    this.note = 'Trimmed mean averages computed over X consecutive official attempts.';
    this.noteZh = '从连续 X 次官方还原中计算的裁剪均值。';
  }

  query(): string { return ''; }

  async toJson(): Promise<StatJson> {
    const metricPanels: MetricPanel[] = [];

    for (const def of AOX_DEFS) {
      const mod = await def.module();
      const StatClass = Object.values(mod).find(v => typeof v === 'function') as
        new () => Statistic;
      const inst = new StatClass();
      const sub = await inst.toJson();

      metricPanels.push({
        id: def.id,
        labelEn: def.label,
        labelZh: sub.titleZh || def.label,
        panels: sub.panels || [],
      });
    }

    return {
      id: this.id,
      title: this.title,
      titleZh: this.titleZh || this.title,
      ...(this.note ? { note: this.note } : {}),
      ...(this.noteZh ? { noteZh: this.noteZh } : {}),
      header: [],
      metricPanels,
    };
  }
}
